Shared Tree: Persisted Schema Format v2 with persisted metadata support

TommyBrosman opened the pull request:
Adds a new Persisted Schema Format with persisted metadata support. This change does not enable it by default.

# Changes

- Add schema FormatV2
- Register FormatV2 with schema codec
- Update codec.spec to cover FormatV2 as well as FormatV1
- Add schema v2 snapshots
- Persisted metadata object in schema
- FormatV2-specific tests
- Documentation

# Notes for reviewers

## Todo

- [x] Changeset description
- [x] simple-tree changes to expose persisted metadata
- [x] Tests for simple-tree layer
- [x] Expose new APIs via alpha types
- [x] Write persisted node and field metadata in the codec
- [ ] Clone and freeze the metadata trait passed to simple-tree

## Related PRs

- Previous PR updating the schema and schema change codecs: https://github.com/microsoft/FluidFramework/pull/24479
- Refactor: Rename CodecFormatVersion to CodecVersion: https://github.com/microsoft/FluidFramework/pull/24666
- Feature flag to turn on persisted metadata: (future PR)

TommyBrosman committed:
Initial change.
- Added in-memory schema format v2.
- Implemented the schema format v2 codec.
- Updated test utils that previously assumed there was only one format.
- Added snapshots for many of the v2 tests.

TommyBrosman committed:
persistedMetadata is now backed by an object instead of a string.

TommyBrosman committed:
Updated codec.spec.ts to cover FormatV2.

TommyBrosman commented on packages/dds/tree/src/codec/codec.ts:
```diff
@@ -352,6 +352,8 @@ export enum FluidClientVersion {
 	/** Fluid Framework Client 2.3 and newer. */
 	v2_3 = "v2_3",
+	/** Fluid Framework Client 2.4 and newer. */
+	v2_4 = "v2_4",
```
Do I need this entry? Should I just add v2_5 instead since 2.40 has already been released?

CraigMacomber replied:
we are missing a ton of entries. 4 is much smaller than 40 ;). You can find a list here: https://www.npmjs.com/package/fluid-framework?activeTab=versions . Every minor should get an entry.

CraigMacomber replied:
Note that extending this enum to have all the entries is pretty unrelated to the other stuff in this PR, and is technically an API change, so I recommend do it in a stand alone PR with a changeset.

TommyBrosman replied:
I'm going to leave release-related changes out of this PR. I think I want to use feature flags. I'll undo this change for now.

TommyBrosman replied:
Removed.

TommyBrosman commented on packages/dds/tree/src/feature-libraries/schema-index/codec.ts:
```diff
@@ -108,7 +127,27 @@ function encodeRepoV1(repo: TreeStoredSchema): FormatV1 {
+			enumerable: true,
+			configurable: true,
+			writable: true,
+			value: schema.encode(),
```
TODO: Make this encode schema v2.

TommyBrosman committed:
Removed extraneous whitespace change.

TommyBrosman committed:
- Changed TreeNodeStoredSchema implementations to take the write version as a parameter.
- Fixed naming of schema index formats.
- Changed schemaChangeFormat to be schema format-agnostic.

TommyBrosman committed:
- Changed the stored schema implementation to handle multiple schema formats.
- Moved SchemaCodecVersion.

TommyBrosman commented on packages/dds/tree/src/core/schema-stored/formatV2.ts:
```diff
@@ -0,0 +1,108 @@
+/*!
```
I'd like feedback on how much of this file should reuse formatV1.

CraigMacomber replied:
I'd recommend only adding things which have changed. I know `FieldKey`, `TreeNodeSchemaIdentifier`, `FieldKindIdentifier` haven't changed for example.

Since the format for v1 will never change what its compatible with, reusing things from it is even lower risk than reusing code in most other cases: here both formats have the exact same stability requirements. I'd recommend reusing everything you can. Doing so will make the review much easier (as only changed code will be in here to review) and easier to maintain across things like TypeBox APi changes ad we will simply have less code to support.

TommyBrosman replied:
Done.

TommyBrosman committed:
Merge branch 'main' into metadata-schema-3

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/formatV2.ts:
```diff
@@ -0,0 +1,108 @@
+	JsonCompatibleReadOnlySchema,
+} from "../../util/index.js";
+
+export const version = 2 as const;
```
do we actually use this any more? I think the new versioning logic and enum have replaced it right?

TommyBrosman replied:
It is referenced in the codec, which can just use the SchemaCodecVersion enum. I can change it.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -9,13 +9,26 @@ import { type MakeNominal, brand, invertMap } from "../../util/index.js";
 
+export enum SchemaCodecVersion {
+	v1 = 1,
+	// Adds persisted metadata to the schema.
```
This is a great comment: make it more useful by using doc comment not a line comment so it will be supported by tooling like IntelliSense.

alexvy86 replied:
I'd even be a bit more specific, "to the node schema and field schema"?

TommyBrosman replied:
Done.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -119,6 +132,13 @@ export interface TreeFieldStoredSchema {
+	 * @remarks
+	 * Discarded when encoding to {@link SchemaFormatVersion.V1}.
+	 */
+	persistedMetadata?: PersistedMetadataFormat;
```
Cases which discard metadata are important (thanks for documenting the main one in the doc comment).

We can make such discarding of metadata more explicit if we mark this as a require field. This will also help find any places which need to be updated to preserve metadata.
```suggestion
	persistedMetadata: PersistedMetadataFormat | undefined;
```

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -160,7 +181,9 @@ export abstract class TreeNodeStoredSchema {
 	 * and is runtime validated by the codec.
 	 */
-	public abstract encode(): TreeNodeSchemaDataFormat;
+	public abstract encode(
```
doing this with a single method and an enum makes every implementation switch over the enum, and every caller have to cast the result.

It seems like it would be more type safe and about the same amount of code or less to have two methods.

That will also scale better to if we add another format in the future which doesn't impact how nodes schema re stored (ex: if format 3 only changes the top level object, having an enum here would make all the subclasses implement it even if its the same as v2)

TommyBrosman replied:
I was debating whether to split it. I think that approach makes sense.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -301,11 +356,20 @@ function decodeValueSchema(inMemory: PersistedValueSchema): ValueSchema {
+
+export function encodeFieldSchemaV2(schema: TreeFieldStoredSchema): FieldSchemaFormatV2 {
 	return {
 		kind: schema.kind,
 		// Types are sorted by identifier to improve stability of persisted data to increase chance of schema blob reuse.
 		types: [...schema.types].sort(),
```
Might as well make it clear v2 is just v1 plus metadata
```suggestion
		...encodeFieldSchemaV1(schema),
```

TommyBrosman replied:
That's clever. I like it.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/formatV2.ts:
```diff
@@ -0,0 +1,108 @@
+		/**
+		 * Persisted metadata for this node.
+		 */
+		persistedMetadata: PersistedMetadataFormat,
```
Annoyingly, we persist this data as JSON, so the names you give things matter for size. Since everything in here persisted, I think we should shorten the name to save document summary size and schema upgrade op size.
```suggestion
		metadata: PersistedMetadataFormat,
```

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/formatV2.ts:
```diff
@@ -0,0 +1,108 @@
+		/**
+		 * Persisted metadata for this node.
+		 */
+		persistedMetadata: PersistedMetadataFormat,
```
putting the metadata here, as a member of this discriminated union doesn't make sense. I don't think we want to store a node schema or metadata  anywhere. Either add it to the thing holding the union, or to the members of the union which should have it, See the linked type and its docs in the doc comment above, `DiscriminatedUnionDispatcher`

TommyBrosman replied:
Fixed.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/formatV2.ts:
```diff
@@ -0,0 +1,108 @@
+		 */
+		leaf: Type.Optional(Type.Enum(PersistedValueSchema)),
+		/**
+		 * Persisted metadata for this node.
```
this is a schema for nodes, not a node. As both are things that we encode: it is important to be clear about it.

TommyBrosman committed:
Apply suggestions from code review

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
- Switched to using the SchemaCodecVersion in public APIs. Eventually we want something more intuitive (e.g., minimum client version).
- Split the encode method for TreeNodeStoredSchema.

TommyBrosman committed:
Merge branch 'metadata-schema-3' of https://github.com/TommyBrosman/FluidFramework into metadata-schema-3

TommyBrosman committed:
- Removed the schema version constant and replaced it with SchemaCodecVersion.
- Changed schema-stored/FormatV2 to only define the new elements.
- Updated all instantiations of TreeFieldStoredSchema so that persistedMetadata is specified.
- HACK: importCompressed needs to handle both schema formats. A few of the roundtrip JSON tests are failing.

TommyBrosman commented on packages/framework/fluid-framework/api-report/fluid-framework.alpha.api.md:
```diff
@@ -281,7 +281,8 @@ export enum FluidClientVersion {
     v2_2 = "v2_2",
-    v2_3 = "v2_3"
+    v2_3 = "v2_3",
+    v2_4 = "v2_4"
```
TODO: Regenerate.

TommyBrosman commented on packages/dds/tree/src/shared-tree/treeAlpha.ts:
```diff
@@ -316,7 +315,8 @@ export const TreeAlpha: TreeAlpha = {
 		const content: ViewContent = {
-			schema: extractPersistedSchema(config, currentVersion),
+			// TODO
+			schema: extractPersistedSchema(config, SchemaCodecVersion.v2),
```
We need to determine which codec version to use when extracting the schema. In the previous world this used the client version.

TommyBrosman replied:
We discussed this offline. This code should use the v1 codec for now.

alexvy86 commented on packages/dds/tree/src/core/schema-stored/formatV2.ts:
```diff
@@ -0,0 +1,48 @@
+		/**
+		 * Persisted metadata for the schema.
+		 */
+		metadata: PersistedMetadataFormat,
```
Is it deliberate that it's `metadata` in node schemas, and `persistedMetadata` in field schemas?

TommyBrosman replied:
Should be metadata in both. I can change it.

alexvy86 commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -160,7 +181,15 @@ export abstract class TreeNodeStoredSchema {
+	public abstract encodeV1(): TreeNodeSchemaDataFormatV1;
+
+	/**
+	 * @privateRemarks
+	 * Returns TreeNodeSchemaDataFormat.
+	 * This is uses an opaque type to avoid leaking these types out of the package,
+	 * and is runtime validated by the codec.
```
Seems like this and encodeV1 need an update/rewrite. The first line feels a bit redundant since the return type is clear in the API signature; is something like "the return type is an opaque type to avoid leaking <WHICH TYPES?> out of the package..."?

TommyBrosman replied:
I removed the comment. It is no longer relevant.

alexvy86 commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -185,18 +214,41 @@ export class ObjectNodeStoredSchema extends TreeNodeStoredSchema {
+		// This makes comparing schema easier, and ensures chunk reuse for schema summaries isn't needlessly broken.
+		for (const key of [...this.objectNodeFields.keys()].sort()) {
+			const value = encodeFieldSchemaV1(
+				this.objectNodeFields.get(key) ?? fail(0xae6 /* missing field */),
```
We shouldn't duplicate assert/fail hex codes. The release process will take care of replacing the string with a new one.
```suggestion
				this.objectNodeFields.get(key) ?? fail("missing field"),
```

alexvy86 commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -185,18 +214,41 @@ export class ObjectNodeStoredSchema extends TreeNodeStoredSchema {
 		// This makes comparing schema easier, and ensures chunk reuse for schema summaries isn't needlessly broken.
 		for (const key of [...this.objectNodeFields.keys()].sort()) {
+			const value = encodeFieldSchemaV1(
+				this.objectNodeFields.get(key) ?? fail(0xae6 /* missing field */),
```
Curious why the update to the hex code or if it was a mistake. Since this is the same codepath as the only one before the PR, I think keeping the same hex code makes sense here.
```suggestion
				this.objectNodeFields.get(key) ?? fail(0xae7 /* missing field */),
```

alexvy86 commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -185,18 +214,41 @@ export class ObjectNodeStoredSchema extends TreeNodeStoredSchema {
+		};
+	}
+
+	public override encodeV2(): TreeNodeSchemaDataFormatV2 {
```
Encoding of metadata for node schema is still pending, I assume?

TommyBrosman replied:
Should work now. I updated encodeV2 so that it uses the same DiscrimiatedUnionDispatcher as encodeV1 and then wraps it in the parent object containing the metadata.

TommyBrosman committed:
Minor: reverted changes to FluidClientVersion utils.

TommyBrosman committed:
Switched back to using min client version for importCompressed.

TommyBrosman committed:
Minor: reverted accidental change.

TommyBrosman committed:
Apply suggestions from code review

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

TommyBrosman committed:
Made persisted metadata schema field naming consistent.

TommyBrosman committed:
- Created a separate storedSchemaDecodeDispatcher for v2 schemas.
- Made importCompressed always use a v1 schema when encoding.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/schema-index/codec.ts:
```diff
@@ -41,8 +39,17 @@ export enum SchemaCodecVersion {
 ): SchemaCodecVersion {
-	// Only one version of the schema codec is currently supported.
-	return SchemaCodecVersion.v1;
+	switch (clientVersion) {
```
Since you are planning to just enable this with a feature flag and not use FluidClientVersion, you don't need to change this. You also shouldn't make this change until you have actually shipped working read support for the format, and updated FluidClientVersion to actually contain the right versions (something you punted so this change should not be done at all as part of this work).

Leave everything using this writing the old v1 format.


Also fluid framework 2.4.0 shipped 7 months ago, and did not have support for reading your new format: do not enable writing that format when trying to be compatible with such versions.

Do not confuse fluid framework 2.4 with 2.40.

TommyBrosman replied:
I was able to remove the switch statement.

CraigMacomber commented on packages/dds/tree/src/test/feature-libraries/schema-index/codecUtil.ts:
```diff
@@ -3,16 +3,36 @@
+ * @param schemaFormat - The schema format version.
+ * @returns The Fluid client version that supports the provided schema format.
+ */
+export function schemaFormatToClientVersion(
```
I don't think there is a use case for this conversion.

TommyBrosman replied:
Agreed. I removed it.

CraigMacomber commented on packages/dds/tree/src/test/feature-libraries/schema-index/codecUtil.ts:
```diff
@@ -3,16 +3,34 @@
+		case SchemaCodecVersion.v1:
+			return FluidClientVersion.v2_0;
+		default:
+			assert(false, `Unsupported schema format: ${schemaFormat}`);
```
Yes this should use unreachableCase. Also any case wanting to do assert(false in the tree code should use fail (and not dynamically derive an assert message).

TommyBrosman replied:
I was able to remove the function.

TommyBrosman committed:
- Refactored the node kind dispatch.
- Added better test data to one of the codec tests for the v2 schema format.

TommyBrosman committed:
Updated snapshots.

TommyBrosman committed:
- Removed minimum client version changes. Equivalent changes will be added in another PR.
- Updated a comment.

TommyBrosman commented on packages/dds/tree/src/test/feature-libraries/schema-index/codecUtil.ts:
```diff
@@ -3,16 +3,34 @@
 ).filter((format) => format !== undefined) as SchemaCodecVersion[];
+
+/**
+ * Convert a schema version to the minimum Fluid client version supporting that format.
+ * @param schemaFormat - The schema format version.
+ * @returns The Fluid client version that supports the provided schema format.
+ */
+export function schemaFormatToClientVersion(
+	schemaFormat: SchemaCodecVersion,
```
I'm glad you're asking about this. I didn't really like the name I came up with, I was just trying to align it with some of the existing language around describing codec versions and write formats. Perhaps `SchemaVersion` would be better?

alexvy86 replied:
I would lean towards that, barring @CraigMacomber or @Abe27342 pointing out flaws or other potential confusion with it.

TommyBrosman replied:
@Abe27342 and I talked, `SchemaVersion` makes sense. I will also make a pass over the doc comment when I change it.

TommyBrosman replied:
Now in #24666

TommyBrosman committed:
Removed more minimumClientVersion glue.

TommyBrosman commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -224,10 +264,23 @@ export class MapNodeStoredSchema extends TreeNodeStoredSchema {
+				};
+			}
+			default:
+				fail(`Cannot decode schema version ${schemaWriteVersion}`);
```
I split these into separate functions, so the switch statement is gone.

TommyBrosman commented on packages/dds/tree/src/test/utils.ts:
```diff
@@ -630,17 +630,13 @@ export function validateTree(tree: ITreeCheckout, expected: JsonableTree[]): voi
 	assert.deepEqual(actual, expected);
 }
 
+// If you are adding a new schema format, consider changing the encoding format used for this codec, given
+// that equality of two schemas in tests is achieved by deep-comparing their persisted representations.
+// If the newer format is a superset of the previous format, it can be safely used for comparisons. This is the
+// case with schema format v2.
```
The comment is specifically referring to the schemaCodec instance below, not all schema codecs. The "this" is kind of weird in that sentence. I'd have to document where `schemaCodec` is in `SchemaCodecVersion`'s doc comment otherwise, and that seems confusing. I think the wording could be better though. Any suggestions?

alexvy86 replied:
Hmm fair, having to point here if we move the doc seems clunky. Then just keep the assert and keep the comment there? So there's _something_ that tells the next person that they might want to make changes here when we add a new schema format? Other than that, personally I don't find the wording particularly confusing.

TommyBrosman committed:
Updated API files.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/schema-index/formatV2.ts:
```diff
@@ -0,0 +1,31 @@
+		version: Type.Literal(SchemaCodecVersion.v2),
+		nodes: Type.Record(Type.String(), schemaFormatV2.TreeNodeSchemaDataFormat),
+		root: schemaFormatV2.FieldSchemaFormat,
+		metadata: schemaFormatV2.PersistedMetadataFormat,
```
Do we have a use-case for this top level metadata? We already have metadata on the root field, and on every node, and on their fields. That seem like enough? I don't see any addition to TreeStoredSchema to allow this data to be stored on it either.

TommyBrosman committed:
Rename: SchemaCodecVersion -> SchemaVersion.

CraigMacomber commented:
In the context of SharedTree, "Schema" without any specific qualifiers (meaning not "Stored Schema" or "Persisted Schema") refers to the "View Schema" the user facing API and objects for defining schema (TreeNodeSchema, SchemaFactory etc).

As this PR makes not changes at that layer, I don't think "Schema v2 with persisted metadata support" is a great name for this.

I think "Persisted Schema Format v2 with persisted metadata support" would be better.

Similar changes should be made to the PR description as well.

CraigMacomber commented:
This PR does not meet the criteria in https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines (also linked from the default PR description via the [template](https://github.com/microsoft/FluidFramework/blob/main/.github/pull_request_template.md)) to be marked as non-draft.

I think the status of this PR and the kind of feedback desired could be better communicated by marking it as draft until those criteria are fulfilled.

TommyBrosman committed:
Merge branch 'main' into metadata-schema-3

TommyBrosman committed:
Removed top-level persisted metadata.

TommyBrosman committed:
Fixed comments.

TommyBrosman committed:
Exposed persistedMetadata as a JsonCompatibleReadOnlyObject.

TommyBrosman committed:
Changeset.

TommyBrosman committed:
Wired up toStoredSchema. Still needs tests.

TommyBrosman commented on packages/dds/tree/src/simple-tree/toStoredSchema.ts:
```diff
@@ -110,13 +110,18 @@ export function getStoredSchema(schema: SimpleNodeSchema): TreeNodeStoredSchema
+			return new MapNodeStoredSchema({
+				kind: FieldKinds.optional.identifier,
+				types,
+				persistedMetadata: undefined,
```
Done, but needs tests.

TommyBrosman committed:
- Updated an old snapshot that didn't include the metadata field.
- Added tests calling schemaFactoryAlpha.

CraigMacomber commented on .changeset/brown-dingos-switch.md:
```diff
@@ -0,0 +1,7 @@
+---
+"@fluidframework/tree": minor
+"__section": feature
+---
```
Users of tree if they are using our public API surface should import it from the reexports in fluid-framework so pretty much all tree changes (except those only impacting legacy APIs) should be in both. Also use the tree section (I'm not super clear why we have that section, but it should be used when making tree changes)

```suggestion
---
"@fluidframework/tree": minor
"fluid-framework": minor
"__section": tree
---
```

CraigMacomber commented on .changeset/brown-dingos-switch.md:
```diff
@@ -0,0 +1,7 @@
+"@fluidframework/tree": minor
+"__section": feature
+---
+Added a v2 Shared Tree schema with support for persisted metadata.
```
This is used as a heading, and thus should not end in a period.

```suggestion
Added a v2 Shared Tree schema with support for persisted metadata
```

TommyBrosman commented on packages/dds/tree/src/feature-libraries/schema-edits/schemaChangeCodecs.ts:
```diff
@@ -45,21 +48,23 @@ export function makeSchemaChangeCodec(
 /**
  * Compose the v1 schema change codec.
  * @param options - The codec options.
+ * @param schemaWriteVersion - The schema write version.
```
This is the convention I discussed with @Abe27342. Essentially this is "v1 + some stuff as specified by the parameters." I'm trying to keep it consistent with existing codecs.

CraigMacomber commented on packages/dds/tree/src/core/schema-stored/schema.ts:
```diff
@@ -164,12 +184,14 @@ export abstract class TreeNodeStoredSchema {
+	/**
+	 * Encode in the v2 schema format.
+	 */
+	public abstract encodeV2(): TreeNodeSchemaDataFormatV2;
```
The return type of this method includes the metadata, but the implementations do not and rely on the caller to provide it.

I think it would be much more type safe to instead have this return V2's TreeNodeSchemaUnionFormat if you want the caller to add the metadata.

CraigMacomber commented on packages/dds/tree/src/test/testTrees.ts:
```diff
@@ -318,6 +326,7 @@ export const testTrees: readonly TestTree[] = [
 			rootFieldSchema: {
 				kind: FieldKinds.required.identifier,
 				types: new Set([allTheFieldsName]),
+				persistedMetadata: undefined,
```
Given that test tree now have persistedMetadata, it seems like they should include a case which uses it.

TommyBrosman commented on packages/dds/tree/src/simple-tree/toStoredSchema.ts:
```diff
@@ -89,7 +89,7 @@ export function convertField(schema: SimpleFieldSchema): TreeFieldStoredSchema {
 		convertFieldKind.get(schema.kind)?.identifier ?? fail(0xae3 /* Invalid field kind */);
 	const types: TreeTypeSet = schema.allowedTypesIdentifiers as TreeTypeSet;
-	return { kind, types };
+	return { kind, types, persistedMetadata: undefined };
```
Here's the comment from the declaration on `TreeFieldStoredSchema`:
```
	/**
	 * Portion of the metadata which can be persisted.
	 * @remarks
	 * Discarded when encoding to {@link SchemaFormatVersion.V1}.
	 */
	readonly persistedMetadata: PersistedMetadataFormat | undefined;
```

I think that is still unambiguous if I change it to `metadata`.

TommyBrosman committed:
Refactor: persistedMetadata -> metadata on persisted types.

TommyBrosman committed:
Apply suggestions from code review

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
- Updated the changeset description
- Moved JsonCompatibleReadOnly and JsonCompatibleReadOnlyObject to alpha API status so they can be used in schemas
- Implemented field factory methods and added test coverage
- Added requiredRecursive to schemaFactoryAlpha (not sure if it should be part of this change though)

TommyBrosman commented on packages/dds/tree/src/simple-tree/api/schemaFactoryAlpha.ts:
```diff
@@ -155,16 +163,67 @@ export class SchemaFactoryAlpha<
+	/**
+	 * {@inheritDoc SchemaStatics.optionalRecursive}
+	 */
+	public static override readonly requiredRecursive = schemaStatics.requiredRecursive;
```
Should this API change be part of this PR? I noticed we were missing the `required` variety and added the recursive alpha variant below.

CraigMacomber replied:
No. Changes which have no need to be part of a larger PR and fix something should be merged independently so they can land sooner and make review of both it and the other PR easier.

Also note that I believe some of these changes are unnecessary. I'm not sure what this optionalRecursiveAlpha method is supposed to provide that's not covered by the existing optionalRecursive overridden static property.

If it is suppose to provide something new, don't duplicate its implementation and signature like this: just assign from the property from schemaStatics like the  static optionalRecursive does.

TommyBrosman committed:
Reverted unnecessary change.

TommyBrosman committed:
Wired up node schema metadata persistence. Currently errors on tests involving FormatV2.

TommyBrosman committed:
Reverted unneeded change.

TommyBrosman committed:
Removed tests for the alpha API I removed in a previous revision.

TommyBrosman committed:
This revision adds simple-tree persistence for node and field schema metadata.

- Fixed a bug where FormatV2 was including FormatV1 fields.
- persistedMetadata is now only written when non-undefined.
- simple-tree implementations for individual schema kinds (arrayNode, mapNode, objectNode) now pass persistedMetadata down from the outer factory calls.
- Introduced alpha APIs for simple-tree in places where we need to pass persistedMetadata but don't want to pollute the public API.
- Updated tests to pass persistedMetadata when constructing SimpleNodeSchemas.
- Wrote simple tests for node and field persisted metadata.

TommyBrosman commented on packages/dds/tree/src/test/simple-tree/api/schemaFactory.spec.ts:
```diff
@@ -414,6 +414,56 @@ describe("schemaFactory", () => {
+			const factory = new SchemaFactoryAlpha("");
+
+			const fooMetadata = {
+				persistedMetadata: { "a": 2 },
```
Doesn't need to be `persistedMetadata` here.

TommyBrosman commented on packages/dds/tree/src/test/simple-tree/api/schemaFactory.spec.ts:
```diff
@@ -414,6 +414,56 @@ describe("schemaFactory", () => {
+				baz: schemaFactory.optionalAlpha(schemaFactory.string, {
+					persistedMetadata: fooMetadata,
+				}),
+				qux: schemaFactory.optionalRecursiveAlpha(
```
Needs to conform to example usages of the recursive field APIs more closely.

TommyBrosman committed:
- Fixed missing persistedMetadata on fields. toStoredSchema.ts and sharedTree.ts weren't copying them in all cases where the needed to, and I was missing SimpleFieldSchema.persistedMetadata.
- Added the v5 Shared Tree format and associated snapshots.
- Fixed FieldSchemaAlpha.persistedMetadata so it won't return an empty object if the data is undefined.

TommyBrosman committed:
Merge branch 'main' into metadata-schema-3

TommyBrosman commented:
Closing. The changes from this PR ended up in:
- #24759
- #24812
